### ISSUE TITLE
Deprecate twisted.cred.credentials.UsernameHashedPassword

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8368]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)| @mesozoic reported|
|-|-|
|Trac ID|trac#8368|
|Type|task|
|Created|2016-05-29 21:06:38Z|
|Branch|https://github.com/twisted/twisted/tree/8368-deprecate-usernamehashedpassword|

See [#3648](https://github.com/twisted/twisted/issues/3648) for historical context on why we're deprecating this class.

<details><summary>Searchable metadata</summary>

```
trac-id__8368 8368
type__task task
reporter__mesozoic mesozoic
priority__normal normal
milestone__None None
branch__8368_deprecate_usernamehashedpassword 8368-deprecate-usernamehashedpassword
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1464555998589539 1464555998589539
changetime__1596315532363470 1596315532363470
version__None None
owner__twm twm

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_1"></a>|@mesozoic removed owner|
|-|-|

Responding to feedback from ticket:3648#comment:22 so that I can finally get this off my plate and pick up something interesting.

> 15.2.0 was already released ... the branch should be updated to 15.3.0

Updated to 16.3.0 :)

> From what I can see, the latest proposed changes does not fix the bug but just deprecate the code. I think that you should either update the ticket description/title or create a new ticket dedicated to deprecation... and close this ticket as wont fix.

Here we are!

> The deprecation code needs a dedicated new test. Check current test suite for test named 'test_deprecation*`

Not sure I understand the value here. We use `callDeprecated` in two separate locations; seems like good enough coverage for the fact that it's deprecated. At the point where those two tests get removed, the class is probably getting removed.

> Maybe the news file should be updated to "twisted.cred.credentials.UsernameHashedPassword is now deprecated."

You got it.

https://github.com/twisted/twisted/pull/84

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @mesozoic|
|-|-|

Hi,

Thanks for the branch.

------

1.Please see our deprecation policy for information about deprecating a class

http://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html#classes

The standard method for deprecating a class is to raise a warning at import time and not each time a class is instantiated.

-------

2. Also, check http://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html#testing-deprecation-code for information about testing the deprecations.

using callDeprecated is existing tests is fine, but we need a new test using flushWarnings to make sure the warning is raised at the right level.

callDeprecated will work even if the warning is raised from a different level

------

I am aware that all the flushWarnings test dance is tedious, but this is the current standard.

If you want to this changed, please raise this issue over the Twisted mailing list.

I can tell that the dedicated flushWarning tests were very useful in the past as callDeprecated is sloppy and the deprecation helpers were not well tested or people were manually using warnings.warn() with a wrong stack level.

------

Please check my comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_4"></a>|@mesozoic removed owner|
|-|-|

Cool, that all makes sense. Thanks for clarifying. I've updated the pull request and kicked off another buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Thanks.

Changes look good.

I think that we need to improve the documentation in the 'testing deprecation code' to make it clear that no warning should be left behind.

The updates for callDeprecated were ok as the test suite should not leave any deprecation code unhandled.

I am not sure why the builders are not reporting these warnings as I see that credentials.UsernameHashedPassword is used in other tests and the warnings are not handled.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @mesozoic|
|-|-|

The last part of the Compatibility/deprecation guide contain info about checking for deprecation warnings


I did

```
$ python -Werror::DeprecationWarning ./bin/trial twisted/cred/test/test_cred.py
```


and it failed.


Please look at updating your branch so that it produce no new warnings.

In the process feel free to update the deprecation documentation based on your current experience.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_7"></a>|@mesozoic removed owner|
|-|-|

OK, I've looked at this, and I think the best thing to do is nuke the test cases that reference this class. `callDeprecated` doesn't work with a class that's a deprecated module attribute; the warning gets raised when our test code references the class, not when calling it. I suspect that may be why the new warnings weren't caught.

I've updated the pull request with more deleted code and kicked off another Buildbot run.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @mesozoic|
|-|-|

Can you please also update the docs/core/development/policy/compatibility-policy.rst file to provide some details about how deprecated classes should be tests?

I would suggest to also send an email over the ML to let other developers know about this method.

-----

In general I am not very happy with removing tests. We might want to deprecate something ... but then during the 1 year deprecation time, we might find security bugs for the deprecated code and we might want to fix them.

Also, during that 1 year of deprecation time, the code is still supported and we should make sure that we don't introduce regressions.

------

I would say that we should find and document a better way.

What do you say if:

* all normal/existing imports are done using something like callDeprecated
* the deprecation test will use reload() py2 / imporlib.reload() py3 ... t.python.compat.reload() to trigger the warning 

-----

Please check my comments and resubmit for review with updated documentation and examples :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

As long as you are importing just the module, the warning is not raised.

I have created [#8478](https://github.com/twisted/twisted/issues/8478) to update the documentation.

----------

Please revert the removed tests.

You can use an import statement like

```
# This is imported in this way to work arround the UsernameHashedPassword
# deprecation warning.
from twisted.cred import credentials
from twisted.cred.credentials import UsernamePassword, IUsernamePassword
from twisted.cred.credentials import IUsernameHashedPassword
```

----------

Why not also deprecate IUsernameHashedPassword ?

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_10"></a>|@mesozoic commented|
|-|-|

OK, I've added a new `getDeprecatedModuleAttribute` helper method for `TestCase`, which will retrieve a module attribute that's been marked for deprecation without triggering the deprecation warning. This allows us to refer to this class in a couple tests without having to do anything funky to suppress import-time warnings.

I chose not to deprecate `IUsernameHashedPassword` because it is used in a number of other places _besides_ UsernameHashedPassword (e.g. DigestedCredentials) and I thought it best to keep the scope of this branch small. Since interfaces get used at the top of a module, I'm not sure how we'd suppress the DeprecationWarnings generated by accessing them.

I'll submit for review as soon as the tests pass.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)|@mesozoic removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @mesozoic|
|-|-|

Thanks for the changes. They look good.

Major comments.

1. In getDeprecatedModuleAttribute the branch for `if len(warningsShown) == 0:` is not covered

2. getDeprecatedModuleAttribute is a public API and we should have a `twisted/topfiles/8368.feature` advertising it. ... trial is in the public API realm

3. Where are the test for getDeprecatedModuleAttribute ? Please consider adding dedicated tests. I think that we are recommending using flushWarnings for the actual deprecation tests since the deprecation helpers have little tests and no way to tell if the are correct or not.

4. We should also update this docstring https://github.com/twisted/twisted/blob/trunk/twisted/python/deprecate.py#L20

Minor comments:

1. test_deprecationUsingFlushWarnings docstring should be updated to reflect that it is no longer the recommended method. 

2. For getDeprecatedModuleAttribute instead of `message` maybe we can just have `replacement` and only the pass replacenet name... as the other string should be standard.

3. I would argue that getDeprecatedModuleAttribute can be merged in  [#8478](https://github.com/twisted/twisted/issues/8478) and once that branch is merged in trunk, we should came back at this ticket. 

4. For IUsernameHashedPassword deprecation, I think that your comment from the Trac ticket would be a good in code comment to describe why are we stuck with that interface.

5. I would argue that we should remove the extensive docstring documentation from https://github.com/twisted/twisted/blob/trunk/twisted/python/deprecate.py and make sure all missing information is in the .rst docs

-----

Please check my comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

https://github.com/twisted/twisted/pull/734

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @rodrigc|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="twm's avatar" src="https://avatars.githubusercontent.com/u/43662?s=50" width="50" height="50">](https://github.com/twm)|@twm removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @twm|
|-|-|

Minor comments but nothing blocking. Land at your leisure. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="twm's avatar" src="https://avatars.githubusercontent.com/u/43662?s=50" width="50" height="50">](https://github.com/twm)<a name="note_20"></a>|@twm set status to `closed`|
|-|-|

In changeset 3b5c938173ba71e956486cdf871443dd0ba05e79
```
#!CommitTicketReference repository="" revision="3b5c938173ba71e956486cdf871443dd0ba05e79"
Merge pull request #734 from twisted/8368-deprecate-usernamehashedpassword

Author: mesozoik, exarkun, twm
Reviewer: glyph
Fixes: ticket:8368

Deprecate UsernameHashedPassword
```
